src/Lecture2.hs

```diff
@@ -243,53 +295,85 @@ Add (Var "x") (Lit 1)
 Add (Var "y") (Add (Var "x") (Lit 10))
 -}
 
-{- | We want to evaluate such expressions. We can associate a value
-with a variable using a list of pairs.
-
-You can use the @lookup@ function to search in this list by a variable name:
-
- * https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:lookup
--}
+-- | We want to evaluate such expressions. We can associate a value
+-- with a variable using a list of pairs.
+--
+-- You can use the @lookup@ function to search in this list by a variable name:
+--
+-- * https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:lookup
 type Variables = [(String, Int)]
 
-{- | Unfortunately, it's not guaranteed that variables in our @Expr@
-data type are present in the given list. So we're going to introduce a
-separate data for possible evaluation errors.
-
-Normally, this would be a sum type with several constructors
-describing all possible errors. But we have only one error in our
-evaluation process.
--}
+-- | Unfortunately, it's not guaranteed that variables in our @Expr@
+-- data type are present in the given list. So we're going to introduce a
+-- separate data for possible evaluation errors.
+--
+-- Normally, this would be a sum type with several constructors
+-- describing all possible errors. But we have only one error in our
+-- evaluation process.
 data EvalError
-    = VariableNotFound String
-    deriving (Show, Eq)
+  = VariableNotFound String
+  deriving (Show, Eq)
 
-{- | Having all this set up, we can finally implement an evaluation function.
-It returns either a successful evaluation result or an error.
--}
+-- | Having all this set up, we can finally implement an evaluation function.
+-- It returns either a successful evaluation result or an error.
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
-
-{- | Compilers also perform optimizations! One of the most common
-optimizations is "Constant Folding". It performs arithmetic operations
-on all constants known during compile time. This way you can write
-more verbose and clear code that works as efficient as its shorter
-version.
-
-For example, if you have an expression:
-
-x + 10 + y + 15 + 20
-
-The result of constant folding can be:
-
-x + y + 45
-
-It also can be:
-
-x + 45 + y
-
-Write a function that takes and expression and performs "Constant
-Folding" optimization on the given expression.
--}
+eval _ (Lit i) = Right i
+eval v (Var s) = case lookup s v of
+  Nothing -> Left $ VariableNotFound s
+  Just i -> Right i
+eval v (Add l r)
+  | isLeft lEval = lEval
+  | isLeft rEval = rEval
+  | otherwise = Right $ fromRight 0 lEval + fromRight 0 rEval
```

Comment:
By combining `isLeft+fromRight` you're using the "Boolean blindness" antipattern. You can use `case-of` instead to write the correct-by-construction function which is a recommended approach.